hedera-node/hedera-admin-service-impl/src/main/java/com/hedera/node/app/service/admin/impl/handlers/FreezeHandler.java

```diff
@@ -31,26 +43,125 @@
  */
 @Singleton
 public class FreezeHandler implements TransactionHandler {
+    // length of the hash of the update file included in the FreezeTransactionBody
+    // used for a quick sanity check that the file hash is not invalid
+    public static final int UPDATE_FILE_HASH_LEN = 48;
+
     @Inject
     public FreezeHandler() {
         // Dagger2
     }
 
     /**
-     * This method is called during the pre-handle workflow.
-     *
-     * <p>Typically, this method validates the {@link TransactionBody} semantically, gathers all
-     * required keys, and warms the cache.
-     *
-     * <p>Please note: the method signature is just a placeholder which is most likely going to
-     * change.
+     * This method is called during the pre-handle workflow for Freeze transactions.
      *
      * @param context the {@link PreHandleContext} which collects all information that will be
      *     passed to {@code handle()}
-     * @throws NullPointerException if one of the arguments is {@code null}
+     * @see <a href="https://hashgraph.github.io/hedera-protobufs/#freeze.proto">Protobuf freeze documentation</a>
      */
-    public void preHandle(@NonNull final PreHandleContext context) {
+    @SuppressWarnings("java:S1874") // disable the warnings for use of deprecated code
+    // it is necessary to check getStartHour, getStartMin, getEndHour, getEndMin, all of which are deprecated
+    // because if any are present then we set a status of INVALID_FREEZE_TRANSACTION_BODY
+    public void preHandle(
+            @NonNull final PreHandleContext context, @NonNull final ReadableSpecialFileStore specialFileStore) {
         requireNonNull(context);
-        throw new UnsupportedOperationException("Not implemented");
+
+        FreezeTransactionBody freezeTxn = context.getTxn().freeze();
+
+        try {
```

Comment:
I would suggest all validation go into a method called `validate` or `preCheck` or similar. And we call that method as first step in preHandle